src/ucp/am/eager_multi.c

```diff
@@ -0,0 +1,146 @@
+        .super.flags         = 0,
+        .first.lane_type     = UCP_LANE_TYPE_AM,
+        .first.tl_cap_flags  = UCT_IFACE_FLAG_AM_BCOPY,
+        .middle.lane_type    = UCP_LANE_TYPE_AM_BW,
+        .middle.tl_cap_flags = UCT_IFACE_FLAG_AM_BCOPY,
+        .max_lanes = init_params->worker->context->config.ext.max_eager_lanes
+    };
+
+    if (init_params->select_param->op_id != UCP_OP_ID_AM_SEND) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return ucp_proto_multi_init(&params, params.super.super.priv,
+                                params.super.super.priv_size);
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_am_fill_middle_footer(ucp_am_mid_ftr_t *ftr, ucp_request_t *req)
+{
+    ftr->msg_id = req->send.msg_proto.message_id;
+    ftr->ep_id  = ucp_send_request_get_ep_remote_id(req);
+}
+
+static UCS_F_ALWAYS_INLINE void
+ucp_am_fill_first_footer(ucp_am_first_ftr_t *ftr, ucp_request_t *req)
+{
+    ucp_am_fill_middle_footer(&ftr->super, req);
+    ftr->total_size = req->send.state.dt_iter.length;
+}
+
+static size_t ucp_am_bcopy_pack_args_first(void *dest, void *arg)
+{
+    ucp_am_hdr_t *hdr                    = dest;
+    ucp_proto_multi_pack_ctx_t *pack_ctx = arg;
+    ucp_request_t *req                   = pack_ctx->req;
+    ucp_am_first_ftr_t *first_ftr;
+    size_t length, max_length;
+
+    ucs_assert(req->send.state.dt_iter.offset == 0);
+
+    ucp_am_fill_header(hdr, req);
+
+    max_length = ucs_min(ucp_am_send_req_total_size(req),
+                         pack_ctx->max_payload);
+    length     = ucp_am_bcopy_pack_data(hdr + 1, req, max_length,
+                                        pack_ctx->next_iter);
+    first_ftr  = UCS_PTR_BYTE_OFFSET(hdr + 1, length);
+
+    ucp_am_fill_first_footer(first_ftr, req);
+
+    return UCP_AM_FIRST_FRAG_META_LEN + length;
+}
+
+static size_t ucp_am_bcopy_pack_args_mid(void *dest, void *arg)
+{
+    ucp_am_mid_hdr_t *hdr                = dest;
+    ucp_proto_multi_pack_ctx_t *pack_ctx = arg;
+    ucp_request_t *req                   = pack_ctx->req;
+    size_t length;
+    ucp_am_mid_ftr_t *mid_ftr;
+
+    /* some amount of data should be packed in the first fragment */
+    ucs_assert(req->send.state.dt_iter.offset > 0);
+
+    hdr->offset = req->send.state.dt_iter.offset;
+    length      = ucp_proto_multi_data_pack(pack_ctx, hdr + 1);
+    mid_ftr     = UCS_PTR_BYTE_OFFSET(hdr + 1, length);
+
+    ucp_am_fill_middle_footer(mid_ftr, req);
+
+    return UCP_AM_MID_FRAG_META_LEN + length;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_eager_am_bcopy_multi_send_func(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        ucp_datatype_iter_t *next_iter)
+{
+    return ucp_proto_eager_bcopy_multi_common_send_func(
+            req, lpriv, next_iter, UCP_AM_ID_FIRST,
+            ucp_am_bcopy_pack_args_first, UCP_AM_FIRST_FRAG_META_LEN,
+            UCP_AM_ID_MIDDLE, ucp_am_bcopy_pack_args_mid,
+            UCP_AM_MID_FRAG_META_LEN);
```

Comment:
Do we need to count `ucp_am_hdr_t` here as well?

Reply:
It is taken into account:
```C
#define UCP_AM_FIRST_FRAG_META_LEN (sizeof(ucp_am_hdr_t) + sizeof(ucp_am_first_ftr_t))
#define UCP_AM_MID_FRAG_META_LEN (sizeof(ucp_am_hdr_t) + sizeof(ucp_am_mid_ftr_t))
```

Reply:
right